r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java

```diff
@@ -698,17 +709,24 @@ public void onError(final Throwable e)
                 LOG.error("Encountered error while invoking error waiter callback", ex);
               }
             }
+
+            // Now after cancelling all the pending tasks, lets make sure to back off on the creation
+            _rateLimiter.incrementPeriod();
```

Comment:
Why is this moved down?

Reply:
call to this method '_rateLimiter.incrementPeriod()' will schedule the next pending task for object creation execution before the current thread get a chance to execute cancelPendingTasks. So by moving this down after cancelPendingTasks - we make sure that the rate limiting tasks only submit new requests.